### PR TITLE
Enabled in-place pod CPU resource updates.

### PR DESCRIPTION
**Pull request by balajismaniam**
- Updated API validation to allow container resource updates.
- Added UpdateContainer() to DockerInterface.
- Updated docker_manger to keep track of containers that require updates.
- Updated docker_manager to update a container appropriately.

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2306,6 +2306,7 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
 	var newContainers []api.Container
 	for ix, container := range mungedPod.Spec.Containers {
 		container.Image = oldPod.Spec.Containers[ix].Image
+		container.Resources = oldPod.Spec.Containers[ix].Resources
```
Comment (ConnorDoyle) [inline pkg/api/validation/validation.go:2309]: In a future iteration, let's expand validation by making sure the inferred QoS class is the same before and after update (if you agree we can make an issue to track it).

File: pkg/kubelet/dockertools/docker_manager.go [hunk 2]
```diff
@@ -558,6 +559,61 @@ func makePortsAndBindings(portMappings []kubecontainer.PortMapping) (map[dockern
...
+			{PathOnHost: "/dev/nvidiactl", PathInContainer: "/dev/nvidiactl", CgroupPermissions: "mrw"},
+			{PathOnHost: "/dev/nvidia-uvm", PathInContainer: "/dev/nvidia-uvm", CgroupPermissions: "mrw"},
+		}
+	}
+
+	uc := dockercontainer.UpdateConfig{
+		Resources: dockercontainer.Resources{
+			Memory:     memoryLimit,
+			MemorySwap: -1,
+			CPUShares:  cpuShares,
```
Comment (ConnorDoyle) [inline pkg/kubelet/dockertools/docker_manager.go:596]: TODO: in future iteration, we should also update the CFS quota if specified and enabled by the kubelet config.

File: pkg/kubelet/dockertools/docker_manager.go [hunk 3]
```diff
@@ -558,6 +559,61 @@ func makePortsAndBindings(portMappings []kubecontainer.PortMapping) (map[dockern
...
+	}
+
+	ref, err := kubecontainer.GenerateContainerRef(pod, container)
+	if err != nil {
+		glog.Errorf("Can't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
+	} else {
+		glog.V(5).Infof("Generating ref for container %s: %#v", container.Name, ref)
+	}
+
+	if err = dm.client.UpdateContainer(string(dockerID), uc); err != nil {
```
Comment (ConnorDoyle) [inline pkg/kubelet/dockertools/docker_manager.go:608]: 👍

**Review by ConnorDoyle — APPROVED** [review]
LGTM.

General comments: Nice work. I wonder if we can factor out the computation of docker resources for reuse in container creation and update. Quite possibly not worth pursuing for PoC though.

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (balajismaniam) [reply pkg/api/validation/validation.go:2309]: Agreed. I will create an issue.

**Pull request merged**